lib/libzfs/libzfs_sendrecv.c

```diff
@@ -5021,14 +5021,15 @@ zfs_receive_one(libzfs_handle_t *hdl, int infd, const char *tosnap,
 			*cp = '@';
 			break;
 		case EINVAL:
-			if (flags->resumable)
+			if (flags->resumable) {
 				zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
 				    "kernel modules must be upgraded to "
 				    "receive this stream."));
-			if (embedded && !raw)
+			} else if (embedded && !raw) {
```

Comment:
is this correct? isn't this supposed to be `(embedded && raw)`?

https://github.com/zfsonlinux/zfs/blob/681a85cb01a9f83b9f742d8fb634edddc68c5679/module/zfs/dmu_recv.c#L398-L400

Reply:
encrypted is not the same as raw. encrypted means that the resulting dataset from the receive was encrypted. Raw means that the encrypted dataset was sent using the raw send feature flag. It is confusing, and a guess at best, but this code should be more accurate than what is there now. I have proposed more general fixes for error reporting in cases like this, which I will hopefully get to some time soon.

Reply:
i also found this:

https://github.com/zfsonlinux/zfs/blob/681a85cb01a9f83b9f742d8fb634edddc68c5679/module/zfs/dmu_recv.c#L573-L580

> embedded data is incompatible with encryption and raw recv

unless i'm missing something we should theoretically handle EINVAL when both `DMU_BACKUP_FEATURE_RAW` and `DMU_BACKUP_FEATURE_EMBED_DATA` are set

Reply:
That should never happen. The flags are mutually exclusive on the send side. There are many places where the recv code can return EINVAL that will be misinterpreted. The idea of this patch is not to correct all of these cases, but simply to make this case (that a few people have hit) function as it was originally intended. We will address the broader problem of badly interpreted error codes in another pull request.